File: java/datastructure/array/leetcode/PascalsTriangle.java
package datastructure.array.leetcode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author roseduan
 * @time 2020/10/30 9:52 上午
 * @description 杨辉三角
 */
public class PascalsTriangle {

    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        for (int i = 0; i < numRows; i++) {
            List<Integer> list = new ArrayList<>();
            for (int j = 0; j < i + 1; j++) {
                if (j == 0 || j == i) {
                    list.add(1);
                } else {
                    list.add(res.get(i - 1).get(j - 1) + res.get(i - 1).get(j));
                }
            }
            res.add(list);
        }
        return res;
    }
}

File: cpp/array/pascals_triangle.cpp
// 杨辉三角

#include "iostream"
#include "vector"

using namespace std;

class Solution {
public:
    vector<vector<int>> generate(int numRows) {
        vector<vector<int>> res = {
            {1},
        };

        for (int i = 1; i < numRows; i++) {
            vector<int> row(i + 1);
            for (int j = 0; j <= i; j++) {
                if (j == 0 || j == i) {
                    row[j] = 1;
                } else {
                    row[j] = res[i - 1][j] + res[i - 1][j - 1];
                }
            }
            res.push_back(row);
        }
        return res;
    }
};
